Client side validation with react-alert (#3050)
section-engineering-education/engineering-education: content/authors/client-side-validation-with-react-alert/index.md

Having a way to check if the data that a user is sending to the webserver of a web application is correct or valid is a feat that JavaScript enables on the web.

> This feature gives frontend developers the ability to create logic that checks the data that users type into the input fields of web apps.

Author: Thank you very much for this feedback. @EspiraMarvin 

I do make use of the Hemingway editor to check my grammar and the readability of my article. But, oftentimes whenever there's a suggestion to make a sentence more readable, I always run out of the best options of sentences that'd suit the scenario.

Thank you once again! I've gone through all the revisions you made, and I've see where I can improve on. I'd go ahead to resolve all of them now.

In this article, we’re going to have a look at how to do this by using an npm package called [react-alert](package link). react-alert makes it easier to render the error messages from the logic that developers create.

> 

Author: Wow! I totally forgot about this.

I'd add the change now.

2. An idea of how conditional statements works in react/JavaScript

3. How npm works and how to install npm packages as dependencies in your projects

4. A little understanding of DOM Manipulation

### Overview

Before React, vanillaJS was the “go-to” solution for client-side validation. One would need to understand the fundamentals of DOM manipulation and event handling.
Since the aim of this validation process is to check if the data or information sent by the user to the server is appropriate by listening to the events that take place on the webpage.

These events range from the click of a button `onClick`, the mouse hovering effect on a particular element on the page `onMouseOver`, the submission of details entered into a webform via the `onSubmit` event and many more that I can’t mention. If you want to know more about events in JavaScript, you should check them out [here](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events).

Performing these tasks (DOM manipulation) in vanillaJS isn’t a big deal because the DOM is available. But in React, the DOM isn’t available a virtual DOM is available instead.

> 

Author: This suggestion is okay. I'd commit it.

Although, manipulating the DOM in React undergoes the same process of doing that in vanillaJS but there are some quirks to it. One of them is how the `document` object isn’t available in React since it is being bundled with nodejs which doesn’t run on the client-side of a web browser.

> 

Author: > Although manipulating the DOM in React is the same process as that of vanilla JS, there are some quirks to it. One is that the `Document` object is not available in React. This is because it is being bundled with Node.js which doesn’t run on the client-side of a web browser.

I feel this section should be: 

Although manipulating the DOM in React **works the same way in** vanilla JS, there are some quirks to it. One is that the `Document` object is not available in React. This is because it is being bundled with Node.js which doesn’t run on the client-side of a web browser.

> 

Author: I've resolved thisat my end

Getting Started
We’ve had a look at what we’d be needing in this article in the first section. To be able to perform client-side validation in React, in the scope of this article, let’s start by installing the necessary `npm` dependencies.

> 

Author: Done

I’d be making use of NextJS to bootstrap my react application because of its simplicity, its folder structure, and many other benefits of using it.

You don’t need to make use of NextJS if you don’t want to, you can decide to make use of [create-react-app](https://create-react-app.dev/) or bootstrapping your react app with [ParcelJS](https://www.digitalocean.com/community/tutorials/how-to-set-up-a-react-project-with-parcel). Feel free to make use of any approach that’d be most convenient for you.

> 

Author: Awesome suggestion!

To get started with NextJS, type the command below into your terminal, it will install all the dependencies that you’d need to create a React application.

`npx create-next-app name-of-your-app`

> 

Author: Alright, I'd take note of this also

> Since this article focuses on the use of “react-alert” for client-side validation in react. We need to also add it to the list of dependencies that’d be used in the project. The command below will handle that for us.

Reviewer: on line 39
Use backticks to highlight keywords(react-alert npm package) i.e
```bash
`react-alert`
```

Also, still on line 39. Replace the word "it" with the "react-alert" package i.e

 We need to also add `react-alert` to the list of dependencies that’d be used in the project.

`npm install react-alert –save-dev`

Foundational knowledge of conditional statements in JavaScript will speed up the process of writing/creating these conditions or checks on the input fields that’d be getting the data that’s being entered into them.

> 

Author: Done

A basic conditional statement is shown below, the `if()` statement. It is the most used one in JavaScript, it checks if the value that is passed as an argument is true, if it isn’t, it returns false.

> ```js

Author: Noted

    alert.error("Please input your email address");
    emailInput.focus();
    errMsg = false;
  } else if (password === "") {
    alert.error("Please provide your password!");
    passwordInput.focus();
    errMsg = false;
  } else if (password.value <= 7) {
    alert.error("Your password must have 8 characters or greater");
    passwordInput.focus();
    errMsg = false;
  } else if (
    typeof password !== "undefined" &&
    typeof confirmPassword !== "undefined"
  ) {
    if (password != confirmPassword) {
      alert.error("Passwords don't match");
      confirmPasswordInput.focus();
      errMsg = false;
    }
  } else {
    alert.success("You've signed up successfully. Proceed to login");
    errMsg = true;
  }
  return errMsg;
};

export default signUpCheck;
```

Wow! I know that might be your reaction right now, but do not fret, We’d have a breakdown of what each snippet does as we move on.

- **Passing form values as props**
  The snippet above is a helper function that performs validation on the type of data that gets into the form field before it is sent to the backend server. The function accepts the `email`, `password`, `confirmPassword`, and `alert` as arguments.

> 

Author: >  The snippet above is a helper function that performs validation on the input form data before the data gets sent to the backend server. The function accepts the `email`, `password`, `confirmPassword`, and `alert` as arguments.

I feel this should be...

 The snippet above is a helper function that performs validation on the input form data before **it** gets sent to the backend server. The function accepts the `email`, `password`, `confirmPassword`, and `alert` as arguments.
 
 To avoid repitions

> 

Author: I've fixed it at my end

```js
let errMsg;

const validateSignUp = (email, password, confirmPassword, alert) => {};
```

We’re passing `alert` as an argument to the function so that we can have access to it when it is imported into the app component. The same thing goes for the other arguments too.

The `errMsg` variable serves as the container that stores and thus displays the error message from the logic. That is why we declared it with the `let` statement, so it can be reassigned to any other value in the code execution process.

- **The conditional statements**

  - The first condition checks if the email field is empty, if it is, an alert error message is rendered in the user interface due to the execution of `alert.error()`

>   The string inside the parenthesis is what will be displayed on the UI, and the browser makes sure that the input field is in focus so that it catches the attention of the user.

Reviewer: ```suggestion
  The `alert.error()` method takes a string as a parameter. The string is what will be displayed on the user interface, and the browser makes sure that the input field is in focus so that it catches the attention of the user.
```

  ```js
  if (!email) {
    alert.error("Please input your email address");
    emailInput.focus();
    errMsg = false;
  }
  ```

>   - The next condition checks if the user enters any password at all. If there’s no value in the form field, an error pops up in the browser, while the next one checks for the validity of the password’s length.

Reviewer: ```suggestion
  - The next condition checks if the user has entered any password at all. If there’s no value in the password form field, an error pops up in the browser. While the next condition checks for the validity of the password’s length.
```

  ```js
  else if (password === "") {
      alert.error("Please provide your password!");
      passwordInput.focus();
      errMsg = false;
  } else if (password.value <= 7) {
      alert.error("Your password must have 8 characters or greater");
      passwordInput.focus();
      errMsg = false;
  }
  ```

>   - This conditional statement checks if there is a match between the first and second passwords. The logic below employs the use of closures in JavaScript to perform this particular validation.

Reviewer: ```suggestion
  - The below conditional statement checks if there is a match between the first and second passwords. The logic employs the use of closures in JavaScript to perform this particular validation.
```

  ```js
  else if (typeof password !== "undefined" && typeof confirmPassword !== "undefined") {
      if (password != confirmPassword) {
          alert.error("Passwords don't match");
          confirmPasswordInput.focus();
          errMsg = false;
      }
  }
  ```

> 
>   It runs a conditional execution on both password fields to see/make sure that both of them are not undefined (i.e that, they have string values in them). If this first condition passes, the next one will run, if not the next one wouldn’t.
> 

Reviewer: ```suggestion

  It runs a conditional execution on both password fields to make sure that both of them are not undefined i.e. They are not empty fields. If this first condition passes, the next one will run, if not the next one would not run.

```

  - If all the conditions stated, have been checked, the alert window comes up with successful text

### Using the validator

Now that we have the validation script set up, and have gotten a grasp of what is going on in the script, it is time to get it inside the app itself. We’d have a look at the step-by-step process of accomplishing that.

> First, we need to have it (the validation script) imported into the App component, alongside the react-alert dependency. For brevity’s sake, I wouldn’t be doing much of an explanation on the React component structure, since it is among the pre-requisites of this article.
> 

Reviewer: ```suggestion
First, we need to import the validation script into the App component, alongside the react-alert dependency. For brevity’s sake, I wouldn’t be doing much of an explanation on the React component structure, since it is among the pre-requisites of this article.

```

                    defaultText="youremail@example.com"
                    value={email}
                    onChange={(e) => setEmail(e.target.value)}
                  />
               </div>
               <div className={style.form__group}>
                <p>Enter Password</p>
                  <input
                    name="password"
                    id="pwd"
                    className={`form-control`}
                    placeholder="password"
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}
                  />
               </div>
               <div className={style.form__group}>
                <p>Confirm Password</p>
                <input
                  name="confirm__pwd"
                  id="pwd__conf"
                  className={`form-control`}
                  placeholder="enter your password again"
                  value={pwdConfirm}
                  onChange={(e) => setPwdConfrim(e.target.value)}
                />
                <button className={`btn-primary`}>Sign up</button>
            </form>
        </div>
    </section>
  );
};

export default withAlert()(SignUp);
```

> Taking a look at the `handleSubmit` function, you’d notice how the arguments from the `signUpCheck` validator are being utilized. In the function, another condition must be met before the form can be submitted.

Reviewer: ```suggestion
Taking a look at the `handleSubmit` function, you’d notice how the arguments from the `signUpCheck` validator are being utilized. In the function, all conditions must be met before the form can be submitted.
```

```js
import React from "react";

import { transitions, positions, Provider as AlertProvider } from "react-alert";

const AlertTemplate = ({ message, options }) => {
  return (
    <div className={options.type === "success" ? "msg-success" : "msg-error"}>
      <p>{message}</p>
    </div>
  );
};

// this component serves as the container
// that holds/displays the error messages due to the validation
// script that runs on all the app components that have an
// input field
const Message = ({ children }) => {
  const options = {
    position: positions.TOP_RIGHT,
    timeout: 3500,
    offset: "0px",
    transition: transitions.SCALE,
  };

  return (
    <AlertProvider template={AlertTemplate} {...options}>
      {children}
    </AlertProvider>
  );
};

export default Message;
```

This snippet below shows a conditional statement but in a ternary syntax. What this translates to, is: when the `options.type` is “success”, add a CSS class of “msg-success”, if not, add a CSS class of “msg-error”.

The “msg-success” gives the alert template/modal a green background while “msg-error” adds a red background to the template.

> Once that is complete, we’d go-ahead to add the template as a parent element in the app component. That can be done by editing the content of `_app.js`

Reviewer: ```suggestion
Once that is complete, we’d go ahead and add the template as a parent element in the app component. This can be done by editing the content of `_app.js` file.
```

```js
import React from “react”
import Message from "../src/utils/Message";
import Head from “next/head”

function App({ Component, pageProps }) {
    return (
      <React.Fragment>
       <Head>
         <link rel="icon" href="#" />
         <meta
           name="viewport"
           content="minimum-scale=1, initial-scale=1, width=device-width"
         />
       </Head>
       <Message>
          <Component {...pageProps} />
       </Message>
      </React.Fragment>
    );
}

export default App;
```

> Wrapping it up

Reviewer: is wrapping it up a heading? If it is use the correct syntax for heading i.e ### Wrapping it up

Author: Yes, It is. I'd make this chnage also

> You’d have noticed that we didn’t dive into the styling of this project, that is because I do not want the article to be too long. You can go on ahead to any styling that befits this project. You can still check out react-alert’s [documentation](https://github.com/schiehll/react-alert) so you have a full understanding of how the package works.

Reviewer: ```suggestion
You have noticed we didn’t dive into the styling of this project, that is because I do not want the article to be too long. You can go ahead and style your project to your own liking.

You can also check out react-alert’s [documentation](https://github.com/schiehll/react-alert) so you get a good understanding of how the package works.
```

### Overview

Before React, vanilla JS was the go-to solution for client-side validation. One would need to understand the fundamentals of DOM manipulation and event handling.

The aim of this validation process is to check if the data sent by the user to the server is appropriate. Event listeners will help us listen to events that take place on the webpage.

These events range from the click of a button `onClick`. The mouse hovering effect on a particular element on the webpage `onMouseOver`. The submission of details entered into an input form using the `onSubmit` event, and much more.

If you want to learn more about events in JavaScript. Visit [Events](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events).
Performing DOM manipulation in vanilla JS isn’t a big deal because the DOM is available. But in React, the DOM isn’t available, a virtual DOM is available instead.

Although manipulating the DOM in React works the same way in vanilla JS, there are some quirks to it. One is that the Document object is not available in React. This is because it is being bundled with Node.js which doesn’t run on the client-side of a web browser

### Getting Started

> 

Author: That's as a result of the code formatting extension I'm using. But, I'd see what I can do about it.

> 

Author: Fixed

2. Or bootstrap your React app with [ParcelJS](https://www.digitalocean.com/community/tutorials/how-to-set-up-a-react-project-with-parcel).
   Feel free to make use of any approach that would be most convenient for you.
   To get started with Next.JS. In your workspace, open the terminal and run the command below.

`npx create-next-app name-of-your-app`
This will install Next.JS all the dependencies needed to create a React application.
Because this article focuses on the use of `react-alert` for client-side validation in react. We need to also add `react-alert` package to the list of dependencies needed in the project. The command below will handle that for us.

```bash
npm install react-alert –save-dev
```

A basic knowledge of how condtional statements operate in JavaScript will speed up the process of writing the validation function(s) on the client-side of your project.

A basic conditional statement is as shown below, the `if` statement. It is one of the most used in JavaScript, it checks if the value passed as an argument is true, if it is not, it returns false.

```javascript
if (username) {
  console.log(username);
}

> ```

Author: Done

                    placeholder="password"
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}
                  />
               </div>
               <div className={style.form__group}>
                <p>Confirm Password</p>
                <input
                  name="confirm__pwd"
                  id="pwd__conf"
                  className={`form-control`}
                  placeholder="enter your password again"
                  value={pwdConfirm}
                  onChange={(e) => setPwdConfrim(e.target.value)}
                />
                <button className={`btn-primary`}>Sign up</button>
            </form>
        </div>
    </section>
  );
};

export default withAlert()(SignUp);
```

Taking a look at the `handleSubmit` function, you’d notice how the arguments from the `signUpCheck` validator are being utilized. In the function, all conditions must be met before the form can be submitted.

```js
const validate = validateSignUp(email, password, pwdConfirm, alert ||" ");

    if (validate) {
      alert.success("You've signed up successfully. Proceed to login");
    }
};
```

- **Setting up the alert template**

> 

Author: Awesome! I'd get right to it

We’d start by creating a component that’d serve as a container for the error messages. In the `utils` folder, we’d add all the logic of the alert-template.

```js
import React from "react";

import { transitions, positions, Provider as AlertProvider } from "react-alert";

const AlertTemplate = ({ message, options }) => {
  return (
    <div className={options.type === "success" ? "msg-success" : "msg-error"}>
      <p>{message}</p>
    </div>
  );
};

// this component serves as the container
// that holds/displays the error messages due to the validation
// script that runs on all the app components that have an
// input field
const Message = ({ children }) => {
  const options = {
    position: positions.TOP_RIGHT,
    timeout: 3500,
    offset: "0px",
    transition: transitions.SCALE,
  };

  return (
    <AlertProvider template={AlertTemplate} {...options}>
      {children}
    </AlertProvider>
  );
};

export default Message;
```

This snippet below shows a conditional statement but in a ternary syntax. What this translates to, is: when the `options.type` is “success”, add a CSS class of “msg-success”, if not, add a CSS class of “msg-error”.

> 

Author: Thank You. I will do this also.

> 

Author: Done

The “msg-success” gives the alert template/modal a green background while “msg-error” adds a red background to the template.

> Once that is complete, we’d go ahead and add the template as a parent element in the app component. This can be done by editing the content of `_app.js` file.
> 

Reviewer: ```suggestion
Let's now go ahead and add the template as a parent element in the app component. This can be done by editing the content of `_app.js` file.

```
